Add support for camera multiplexer boards

6by9 opened the pull request:
Significant rejuggle of the sensor configs.

Tested all single camera variants (including on cam0), and all sensors on at least one port of a mux (Arducam [Doubleplexer](https://www.arducam.com/product/multi-camera-adapter-doubleplexer-stereo-module-v2-raspberry-pi-zero-3-b-4/) and [4 port mux](https://www.arducam.com/product/multi-camera-v2-1-adapter-raspberry-pi/).

6by9 committed:
dtoverlays: Reshuffle image sensor overlays to allow use with muxes

There are third-party CSI-2 muxes available that ideally want to
be configured to have any of the supported image sensors.
That wants the sensor configuration to be split out into fragements
that can be included in the relevant places, and then an overall
overlay that pulls them in.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>

6by9 committed:
defconfigs: Add VIDEO_MUX to all defconfigs

Supports camera multiplexers using the mux framework.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>

6by9 committed:
media: video-mux: Read CSI2 config from FW, and pass to receiver

There is no obligation for all source devices on a video-mux to
require the same bus configuration, so read the configuration
from the sink ports, and relay via get_mbus_config on the source
port.
If the sources support get_mbus_config, then call that first.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>

pelwell commented on arch/arm/boot/dts/overlays/ov2311-overlay.dts:
```diff
@@ -15,31 +15,7 @@
-					};
-				};
-			};
+			#include "ov2311.dtsi"
```
This file is missing - did you forget to include it in the commit?

pelwell commented on arch/arm/boot/dts/overlays/README:
```diff
@@ -686,6 +686,68 @@ Info:   This overlay is now deprecated - see i2c-sensor
+Info:   Configures a 4 port camera multiplexer
+        Note that currently ALL IMX290 modules share a common clock, therefore
+        all modules will need to have the same clock frequency.
+Load:   dtoverlay=camera-mux-2port,<param>=<val>
```
`-4port`.

pelwell commented on arch/arm/boot/dts/overlays/camera-mux-2port.dts:
```diff
@@ -0,0 +1,383 @@
+// SPDX-License-Identifier: GPL-2.0-only
```
This file name is missing the `-overlay`.

pelwell commented on arch/arm/boot/dts/overlays/camera-mux-4port.dts:
```diff
@@ -0,0 +1,667 @@
+// SPDX-License-Identifier: GPL-2.0-only
```
This file name is missing the `-overlay`.

pelwell reviewed (commented):
There are a few major blockers. Once fixed I should be able to review the reset.

pelwell commented on arch/arm/boot/dts/overlays/imx290_327-overlay.dtsi:
```diff
@@ -16,28 +16,7 @@
 			#size-cells = <0>;
```
I can see that they aren't new files, but I'm not that keen on `*-overlay.dtsi` as filenames. And the relationship between this and the new `imx290_327.dtsi` isn't obvious, especially since neither `imx290_327.dtsi` nor `imx477_378.dtsi` appears to be used.

pelwell replied:
Sorry, `imx290_327.dtsi` is used, and `imx477_378.dtsi` isn't even a thing, but the rest of the comment stands.

6by9 commented:
Branch updated to deal with the first 4 blockers.

imx477 and imx378 are identical except for compatible. Likewise imx290 and imx327. 

imx477 and imx378 both pull in imx477_378.dtsi, and that pulls in imx477.dtsi.
The mux overlays also pull in imx477.dtsi

imx290-overlay pulls in imx290_327-overlay.dtsi, and that pulls in imx290_327.dtsi.
Again the mux overlays also pull in imx290_327.dtsi.

Happy to rework filenames as you want.

pelwell commented:
> imx477 and imx378 both pull in imx477_378.dtsi

Show me the lines.

> imx477 and imx378 both pull in imx477_378.dtsi

Ditto.

6by9 commented on arch/arm/boot/dts/overlays/imx477-overlay.dts:
```diff
@@ -5,6 +5,6 @@
 
 #include "imx477_378-overlay.dtsi"
```
imx477-overlay.dts pulls in imx477_378-overlay.dtsi

pelwell replied:
That was just a typo in the comment then - there is no `imx477_378.dtsi`.

I think it's the asymmetry that is most confusing; 477/378 have the following files:
```
imx378-overlay.dts
imx477-overlay.dts
imx477.dtsi
imx477_378-overlay.dtsi
```
while 290/327 have:
```
imx290-overlay.dts
imx290_327.dtsi
imx290_327-overlay.dtsi
```

6by9 replied:
It looks like imx327-overlay.dts never got created in the past, partly as the driver didn't get updated to have the extra compatible string. I'll sort that out.
imx378 did need about 4 different registers set cf imx477, hence the driver did get updated.

I'll rename imx477.dtsi to imx477_378.dtsi so that we at least keep the same distinction over whether it is a file common to multiple devices, or just one.

6by9 commented on arch/arm/boot/dts/overlays/imx477_378-overlay.dtsi:
```diff
@@ -40,31 +40,7 @@
-					};
-				};
-			};
+			#include "imx477.dtsi"
```
imx477_378-overlay.dtsi pulls in imx477.dtsi.

6by9 replied:
Can rename imx477.dtsi to imx477_378.dtsi to make it more consistent.

6by9 reviewed (commented):
I have missed the update to imx378-overlay.dts, but it is the same rename of &imx477 to &cam_node

6by9 committed:
media: i2c: imx290: Add compatible strings for IMX327 and IMX462

IMX327 is the previous generation to IMX290, and supports up to 1080p60
as 10 or 12 bit.
IMX290 adds 1080p120 in 10 bit mode.
IMX462 adds 1080p120 in both 10 and 12 bit modes.

Add compatible strings for all variants.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>

pelwell commented:
Other comments:
* `cam0-ov2311`, `cam1-ov2311`, `cam2-ov2311` and `cam3-ov2311` are undocumented.
* The labels `imx477` and `imx219_0` appear to have been lost.

pelwell commented:
Oh - the macro renaming is tripping up ovmerge. Hmmm - I'll have to think about that.

6by9 commented:
Rename done, and camN-ov2311 documented.

6by9 commented:
Oh, and the extra compatibles for imx290.
Except I forgot to add imx327-overlay.dts and imx462-overlay.dts. Updating again.

6by9 committed:
dtoverlays: Add overlays for Sony IMX327 and IMX462 image sensors

These are currently identical to IMX290, but have slightly
different maximum frame rate capabilities.

Add overlays for the different compatible strings.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>

pelwell commented:
Apart from a bug (imx477_378-overlay.dtsi refers to `vdda-supply` when it should be `VANA-supply`), the rest looks like it should work, but it isn't exactly elegant.

I'll accept the substitution of `cam_node`, `cam_endpoint` and `cam1_clk` as necessary. I'm less keen on the the fact that when applied the overlay includes all the disabled cameras, but apart from bloating the DTB it isn't harmful.

I don't like the global defines - `ADD_COMPATIBLE` and `DISABLE_CAMERA` - both seem easy enough to replace with enables in the individual overlays (`DISABLE_CAMERA`) or remove altogether (`ADD_COMPATIBLE`). I also don't like all the warnings caused by all the camera endpoints declaring what they are connected to when the mux can't refer back to them all - I think both ends should be patched by the wrapper overlay or the parameter overrides that enable that camera. I think this should allow the `csi_ep` substitution to be deleted.

6by9 commented:
Patches deliberately not squashed as there are further issues.

```ADD_COMPATIBLE``` and ```DISABLE_CAMERA``` both removed.

Making the mapping match by not having remote-endpoint in the eg ov9281.dtsi but instead in ov9281-overlay.dts then throws up the error during compile of ov9281-overlay.dts
```
/home/dave/Pi/linux/arch/arm/boot/dts/overlays/ov2311-overlay.dts:18.13-23.4: Warning (graph_port): /fragment@1: graph port node name should be 'port'
/home/dave/Pi/linux/arch/arm/boot/dts/overlays/ov2311-overlay.dts:20.15-22.5: Warning (graph_endpoint): /fragment@1/__overlay__: graph endpoint node name should be 'endpoint'
/home/dave/Pi/linux/arch/arm/boot/dts/overlays/ov2311-overlay.dts:20.15-22.5: Warning (graph_endpoint): /fragment@1/__overlay__: graph connection to node '/fragment@3/__overlay__/port/endpoint' is not bidirectional
/home/dave/Pi/linux/arch/arm/boot/dts/overlays/ov2311-overlay.dts:43.22-46.7: Warning (graph_endpoint): /fragment@3/__overlay__/port/endpoint: graph connection to node '/fragment@2/__overlay__/ov2311@60/port/endpoint' is not bidirectional
/home/dave/Pi/linux/arch/arm/boot/dts/overlays/ov7251-overlay.dts:18.13-23.4: Warning (graph_port): /fragment@1: graph port node name should be 'port'
/home/dave/Pi/linux/arch/arm/boot/dts/overlays/ov7251-overlay.dts:20.15-22.5: Warning (graph_endpoint): /fragment@1/__overlay__: graph endpoint node name should be 'endpoint'
/home/dave/Pi/linux/arch/arm/boot/dts/overlays/ov7251-overlay.dts:20.15-22.5: Warning (graph_endpoint): /fragment@1/__overlay__: graph connection to node '/fragment@3/__overlay__/port/endpoint' is not bidirectional
/home/dave/Pi/linux/arch/arm/boot/dts/overlays/ov7251-overlay.dts:43.22-46.7: Warning (graph_endpoint): /fragment@3/__overlay__/port/endpoint: graph connection to node '/fragment@2/__overlay__/ov7251@60/port/endpoint' is not bidirectional
```
So the compiler appears to be validating each fragment and doesn't do a full lookup of what ```cam_ep``` actually is.
I've checked that ov9281 does still work with that error, and it does.

It would be possible to implement the whole thing with each sensor in a separate fragment, which would avoid the bloat of the DTB. It would mean we can avoid the above errors, but it'll be a large number of fragments.

pelwell commented:
I've pushed single fixup patch that does a number of things:
1. It moves the endpoint settings to either "fixup nodes" (I don't know a better name for the magic "&label {" construction that tells the compiler to apply the patches) in the individual overlays, or to the parameter overrides for the mux overlays.
2. Ensures that all the cameras start out disabled, only enabling them as needed.
3. Regularises the fragment numbering in the mux overlays so that the 2port overlay is almost a subset of the 4port overlay.
4. Moves the setting of the noncontinuous flags to the parameters, because it's more human readable and easier to maintain.

Consider this commit as a bag of bits to be squashed and merged as you wish.

pelwell commented:
And I've got used to the `*-overlay.dtsi` names - they are almost complete overlays, but still need some completion. The other file names changes are good.

6by9 commented:
Thank you. Blasted device tree!

Tested in one configuration of the 4 port, and all seems good. 4 cameras found, all of which stream OK.
Running ov9281, imx219, imx477, and imx327 individually has also worked.

Patches squashed and branch updated.

pelwell commented:
I'm sure that if the people maintaining the tools were overlay users then this would be much easier.

I'm unlikely to find any more problems by inspection/diff - it's time to release it to the beta-testers.

pelwell commented:
The video support in 5.18 has changed sufficiently that a37312d3ac5d939ac54ab87f97366308f6f5b602 no longer applies. The lack of the `V4L2_MBUS_CSI2_CONTINUOUS_CLOCK` flag is easy to fix, but the changes related to the absence of `V4L2_MBUS_CSI2_LANE_MASK` are non-trivial.

pelwell commented:
Is there any chance you could take a look at porting [a37312d ](https://github.com/raspberrypi/linux/commit/a37312d3ac5d939ac54ab87f97366308f6f5b602) to 5.18?

6by9 commented:
> Is there any chance you could take a look at porting [a37312d ](https://github.com/raspberrypi/linux/commit/a37312d3ac5d939ac54ab87f97366308f6f5b602)to 5.18?

Yes, will do.
